lib/luacompletion.c

```diff
@@ -0,0 +1,88 @@
+/*
+* SPDX-FileCopyrightText: (c) 2023-2024 Ring Zero Desenvolvimento de Software LTDA
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/completion.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <lunatik.h>
+
+LUNATIK_OBJECTCHECKER(luacompletion_check, struct completion *);
+
+static int luacompletion_complete(lua_State *L)
+{
+	struct completion *completion = luacompletion_check(L, 1);
+
+	complete(completion);
+	return 0;
+}
+
+static int luacompletion_wait(lua_State *L)
+{
+	struct completion *completion = luacompletion_check(L, 1);
+	lua_Integer timeout = luaL_optinteger(L, 2, MAX_SCHEDULE_TIMEOUT);
+	unsigned long timeout_jiffies = msecs_to_jiffies((unsigned long)timeout);
+	unsigned long ret;
+
+	ret = wait_for_completion_interruptible_timeout(completion, timeout_jiffies);
+	if (ret == 0 || ret == -ERESTARTSYS) {
+		lua_pushnil(L);
+		lua_pushstring(L, ret == 0 ? "timeout" : "interrupted");
+		return 2;
+	}
+	lua_pushboolean(L, true);
+	return 1;
+}
+
+static int luacompletion_new(lua_State *L);
+
+static const luaL_Reg luacompletion_lib[] = {
+	{"new", luacompletion_new},
+	{NULL, NULL}
+};
+
+static const luaL_Reg luacompletion_mt[] = {
+	{"__gc", lunatik_deleteobject},
+	{"complete", luacompletion_complete},
+	{"wait", luacompletion_wait},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luacompletion_class = {
+	.name = "completion",
+	.methods = luacompletion_mt,
+	.sleep = false,
```

Comment:
btw, is it safe to call `wait()` from a softirq? It shouldn't, right? If it isn't, I would add a guard to raise an error.. I've done similar guards in my PR (`lneto_sleep`)